gtsfm/frontend/verifier/oanet.py

```diff
@@ -0,0 +1,188 @@
+        self.default_config["iter_num"] = 1
+        self.default_config["inlier_threshold"] = 1
+        self.default_config_ = SimpleNamespace(**self.default_config)
+
+        self.model = OANet(self.default_config_, self.device)
+
+        checkpoint = torch.load(self.model_path, map_location=self.device)
+        self.model.load_state_dict(checkpoint["state_dict"])
+
+        if self.is_cuda:
+            self.model = self.model.cuda()
+
+        self.model.eval()
+
+        # Run RANSAC after OANet, for additional verification.
+        self.ransac = Ransac(
+            use_intrinsics_in_verification=use_intrinsics_in_verification,
+            estimation_threshold_px=estimation_threshold_px,
+        )
+
+    def verify(
+        self,
+        keypoints_i1: Keypoints,
+        keypoints_i2: Keypoints,
+        match_indices: np.ndarray,
+        camera_intrinsics_i1: Cal3Bundler,
+        camera_intrinsics_i2: Cal3Bundler,
+    ) -> Tuple[Optional[Rot3], Optional[Unit3], np.ndarray, float]:
+        """Performs verification of correspondences between two images to recover the relative pose and indices of
+        verified correspondences.
+
+        Based on OANet's LearnedMatcher class: https://github.com/zjhthu/OANet/blob/master/demo/learnedmatcher.py
+
+        Args:
+            keypoints_i1: detected features in image #i1.
+            keypoints_i2: detected features in image #i2.
+            match_indices: matches as indices of features from both images, of shape (N3, 2), where N3 <= min(N1, N2).
+            camera_intrinsics_i1: intrinsics for image #i1.
+            camera_intrinsics_i2: intrinsics for image #i2.
+
+        Returns:
+            Estimated rotation i2Ri1, or None if it cannot be estimated.
+            Estimated unit translation i2Ui1, or None if it cannot be estimated.
+            Indices of verified correspondences, of shape (N, 2) with N <= N3. These are subset of match_indices.
+            Inlier ratio w.r.t. the estimated model, i.e. #ransac inliers / # putative matches.
+        """
+        if match_indices.shape[0] < self._min_matches:
+            logger.info("[OANet] Not enough correspondences for verification.")
+            return self._failure_result
+
+        if len(keypoints_i1) < self._min_matches:
+            return self._failure_result
+
+        # only matched keypoints are fed into the model.
+        matched_keypoints_i1 = keypoints_i1.coordinates[match_indices[:, 0]]
+        matched_keypoints_i2 = keypoints_i2.coordinates[match_indices[:, 1]]
+
+        # using OANet's custom normalization utility, obtain two (N,2) arrays
+        normalized_keypoints = [
+            torch.from_numpy(self.__normalize_kpts(matched_keypoints_i1).astype(np.float32)).to(self.device),
+            torch.from_numpy(self.__normalize_kpts(matched_keypoints_i2).astype(np.float32)).to(self.device),
+        ]
```

Comment:
Any way we can use calibration to normalize the points? To get rid of distortion for example? Maybe we can do calibration's normalization followed by OANet's normalization